core-tests/shared/src/test/scala/zio/stm/STMSpec.scala

```diff
@@ -1,147 +1,423 @@
+            assert(failedV, isLeft(equalTo(-2)))
+        },
+        suite("Failure must")(
+          testM("rollback full transaction") {
+            for {
+              tvar <- TRef.makeCommit(0)
+              e <- (for {
+                    _ <- tvar.update(_ + 10)
+                    _ <- STM.fail("Error!")
+                  } yield ()).commit.either
+              v <- tvar.get.commit
+            } yield assert(e, isLeft(equalTo("Error!"))) && assert(v, equalTo(0))
+          },
+          testM("be ignored") {
+            for {
+              tvar <- TRef.makeCommit(0)
+              e <- (for {
+                    _ <- tvar.update(_ + 10)
+                    _ <- STM.fail("Error!")
+                  } yield ()).commit.ignore
+              v <- tvar.get.commit
+            } yield assert(e, equalTo(())) && assert(v, equalTo(0))
+          }
+        ),
+        suite("orElse must")(
+          testM("rollback left retry") {
+            for {
+              tvar  <- TRef.makeCommit(0)
+              left  = tvar.update(_ + 100) *> STM.retry
+              right = tvar.update(_ + 100).unit
+              _     <- (left orElse right).commit
+              v     <- tvar.get.commit
+            } yield assert(v, equalTo(100))
+          },
+          testM("rollback left failure") {
+            for {
+              tvar  <- TRef.makeCommit(0)
+              left  = tvar.update(_ + 100) *> STM.fail("Uh oh!")
+              right = tvar.update(_ + 100).unit
+              _     <- (left orElse right).commit
+              v     <- tvar.get.commit
+            } yield assert(v, equalTo(100))
+          },
+          testM("local reset, not global") {
+            for {
+              ref <- TRef.make(0).commit
+              result <- STM.atomically(for {
+                         _       <- ref.set(2)
+                         newVal1 <- ref.get
+                         _       <- STM.partial(throw new RuntimeException).orElse(STM.unit)
+                         newVal2 <- ref.get
+                       } yield (newVal1, newVal2))
+            } yield assert(result, equalTo((2 -> 2)))
+          }
+        ),
+        suite("STM issue 1587") {
+          testM("read only STM shouldn't return partial state of concurrent read-write STM") {
+            def repeat(test: UIO[TestResult], i: Int): UIO[TestResult] =
+              test.flatMap { result =>
+                if (result.isSuccess && i > 0) repeat(test, i - 1)
+                else ZIO.succeed(result)
+              }
+
+            val race = for {
+              r0       <- TRef.makeCommit(0)
+              r1       <- TRef.makeCommit(0)
+              sumFiber <- r0.get.flatMap(v0 => r1.get.map(_ + v0)).commit.fork
+              _        <- r0.update(_ + 1).flatMap(_ => r1.update(_ + 1)).commit
+              sum      <- sumFiber.join
+            } yield assert(sum, equalTo(0) || equalTo(2))
+
+            repeat(race, 10000)
```

Comment:
I have another report that this spec timed out again.
I'll remove this repeat thing and will add barriers inside STM to make sure that:
* updating fiber will update r0 before sum fiber reads it
* sum fiber reads r1 before updating fiber updates it
Without recent fix this would result in sum=1, because of fix, STM will retry sum and result will have to be 2.

Reply:
Yes. I have been seeing timeouts more frequently. I know we don't have this merged yet so that doesn't speak to whether your earlier fix would have been sufficient but it is definitely good to make sure this isn't taking too long.

Reply:
I was too optimistic with this approach. I can make sure that read-only fiber won't start STM before update fiber updates first value but I can't await inside STM, so it's hard to make sure that update thread will suspend until read-only thread is done.